t/unit-tests/u-trailer.c

```diff
@@ -0,0 +1,359 @@
+#define DISABLE_SIGN_COMPARE_WARNINGS
+
+#include "unit-test.h"
+#include "trailer.h"
+
+struct contents {
+	const char *raw;
+	const char *key;
+	const char *val;
+};
+
+struct test_case {
+	const char *name;
+	const char *msg;
+	size_t num_expected;
+	struct contents contents[10];
+};
+
+static void t_trailer_iterator(const char *msg, size_t num_expected,
+			       struct contents *contents)
+{
+	struct trailer_iterator iter;
+	size_t i = 0;
+
+	trailer_iterator_init(&iter, msg);
+	while (trailer_iterator_advance(&iter)) {
+		if (num_expected) {
+			cl_assert_equal_s(iter.raw, contents[i].raw);
+			cl_assert_equal_s(iter.key.buf, contents[i].key);
+			cl_assert_equal_s(iter.val.buf, contents[i].val);
+		}
+		i++;
+	}
+	trailer_iterator_release(&iter);
+
+	cl_assert_equal_i(i, num_expected);
+}
+
+void test_trailer__empty_input(void)
+{
+	struct test_case tc = {
+		"empty input",
+		"",
+		0,
+		{{0}},
+	};
+
+	t_trailer_iterator(tc.msg, tc.num_expected, tc.contents);
+}
```

Comment:
I think there is no real use for `struct test_case` anymore:

- The `name` field is completely unused.
- The other fields are all directly passed to `t_trailer_iterator()`.

So using this testcase here as an example, I think the following would be simpler:

```
void test_trailer__empty_input(void)
{
	struct contents expected_contents = { 0 };
	t_trailer_iterator("", 0, expected_contents);
}
```

Other than that I really like that you've split up the testcases into separate tests. Makes it way easier to follow what is going on and what we're testing right now.